File: src/services/ai/contentDiscovery.ts
// src/services/ai/contentDiscovery.js
// Multi-platform content discovery with LLM scoring and diversity ranking

import { httpPostJson } from './_client.js';
import { AI_SERVICE_TIMEOUT_MS } from '../../config.js';
import { logger } from '../../util/logger.js';

type OrchestratedSearchOptions = {
  platforms?: string[] | null;
  maxResults?: number;
  contentTypes?: string[] | null;
};

/**
 * @typedef {Object} ContentItem
 * @property {string} title - Content title
 * @property {string} url - Content URL
 * @property {string} [snippet] - Content snippet/description
 * @property {string} platform - Platform name (variety.com, youtube, deviantart, etc.)
 * @property {string} content_type - Content type (news, video, discussion, art)
 * @property {string} [published_date] - Publication date
 * @property {number} relevance_score - Relevance score (0-100+)
 * @property {Object} [score_breakdown] - Score breakdown
 * @property {number} score_breakdown.topic - Topic relevance (0-100)
 * @property {number} score_breakdown.quality - Content quality (0-100)
 * @property {number} score_breakdown.recency - Recency bonus (0-20)
 * @property {number} score_breakdown.credibility - Source credibility (0-20)
 * @property {string} [relevance_reasoning] - LLM reasoning for score
 * @property {number} [adjusted_score] - Diversity-adjusted score
 * @property {Object} [attribution] - Attribution (for art)
 * @property {string} [attribution.artist] - Artist name
 * @property {string} [attribution.artist_url] - Artist profile URL
 * @property {string} [attribution.license] - License
 */

/**
 * @typedef {Object} DiscoveryResult
 * @property {string} query - Original search query
 * @property {ContentItem[]} results - Scored and ranked content items
 * @property {number} total_results - Total number of results
 * @property {number} diversity_score - Shannon entropy diversity score (0-100)
 * @property {string[]} platforms_covered - Platforms that returned results
 * @property {number} [duration_ms] - Request duration in milliseconds
 */

/**
 * Orchestrated content discovery across all platforms
 *
 * This function performs:
 * 1. Multi-source aggregation (news, YouTube, Reddit, Twitter, DeviantArt, Tumblr)
 * 2. LLM-powered relevance scoring (topic, quality, recency, credibility)
 * 3. Diversity-aware ranking (platform balancing, duplicate detection)
 *
 * @param {string} query - Search query (e.g., "Elio Pixar movie 2025")
 * @param {Object} options - Discovery options
 * @param {string[]} options.platforms - Platforms to search (default: all)
 *   Options: 'news', 'youtube', 'reddit', 'twitter', 'deviantart', 'tumblr'
 * @param {number} options.maxResults - Maximum total results (default: 20)
 * @param {string[]} options.contentTypes - Desired content types (default: all)
 *   Options: 'news', 'video', 'discussion', 'art'
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 *
 * @example
 * const result = await orchestratedSearch('Elio Pixar movie', {
 *   platforms: ['news', 'youtube', 'deviantart'],
 *   maxResults: 20,
 *   contentTypes: ['news', 'video', 'art']
 * });
 *
 * if (result.ok) {
 *   console.log(`Found ${result.data.total_results} results`);
 *   console.log(`Diversity score: ${result.data.diversity_score}`);
 *   result.data.results.forEach(item => {
 *     console.log(`- ${item.title} (${item.platform}, score: ${item.relevance_score})`);
 *     if (item.attribution) {
 *       console.log(`  Artist: ${item.attribution.artist}`);
 *     }
 *   });
 * }
 */
export async function orchestratedSearch(query: string, {
  platforms = null, // null = all platforms
  maxResults = 20,
  contentTypes = null // null = all content types
}: OrchestratedSearchOptions = {}): Promise<any> {
  const payload = {
    query,
    platforms,
    max_results: maxResults,
    content_types: contentTypes,
  };

  logger.info('[CONTENT-DISCOVERY] Starting orchestrated search', {
    query,
    platforms: platforms || 'all',
    maxResults
  });

  try {
    // Use extended timeout for complex searches
    const timeout = AI_SERVICE_TIMEOUT_MS * 3; // 90 seconds default
    const res = await httpPostJson('/content-discovery/discover', payload, timeout);

    if (res.status >= 400 || !res.json?.ok) {
      logger.warn('[CONTENT-DISCOVERY] Search failed', {
        query,
        status: res.status,
        error: res.json?.error
      });
      return {
        ok: false,
        error: {
          code: 'CONTENT_DISCOVERY_FAILED',
          message: 'Content discovery failed',
          cause: res.json?.error
        }
      };
    }

    const data = res.json.data;

    logger.info('[CONTENT-DISCOVERY] Search completed', {
      query,
      total_results: data.total_results,
      diversity_score: data.diversity_score,
      platforms_covered: data.platforms_covered,
      duration_ms: data.duration_ms
    });

    return {
      ok: true,
      data: {
        query: data.query,
        results: data.results || [],
        total_results: data.total_results || 0,
        diversity_score: data.diversity_score || 0.0,
        platforms_covered: data.platforms_covered || [],
        duration_ms: data.duration_ms,
      }
    };
  } catch (err: any) {
    logger.error('[CONTENT-DISCOVERY] Search error', {
      query,
      error: err.message
    });
    return {
      ok: false,
      error: {
        code: 'DEPENDENCY_UNAVAILABLE',
        message: 'Content discovery service unavailable',
        cause: err
      }
    };
  }
}

/**
 * Search specific platforms (convenience wrapper)
 * @param {string} query - Search query
 * @param {string[]} platforms - Platforms to search
 * @param {number} maxResults - Maximum results
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 */
export async function searchPlatforms(query: string, platforms: string[], maxResults = 20): Promise<any> {
  return orchestratedSearch(query, { platforms, maxResults });
}

/**
 * Search for news only
 * @param {string} query - Search query
 * @param {number} maxResults - Maximum results
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 */
export async function searchNews(query: string, maxResults = 10): Promise<any> {
  return orchestratedSearch(query, {
    platforms: ['news'],
    contentTypes: ['news'],
    maxResults
  });
}

/**
 * Search for videos only (YouTube)
 * @param {string} query - Search query
 * @param {number} maxResults - Maximum results
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 */
export async function searchVideos(query: string, maxResults = 10): Promise<any> {
  return orchestratedSearch(query, {
    platforms: ['youtube'],
    contentTypes: ['video'],
    maxResults
  });
}

/**
 * Search for discussions (Reddit, Twitter)
 * @param {string} query - Search query
 * @param {number} maxResults - Maximum results
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 */
export async function searchDiscussions(query: string, maxResults = 10): Promise<any> {
  return orchestratedSearch(query, {
    platforms: ['reddit', 'twitter'],
    contentTypes: ['discussion'],
    maxResults
  });
}

/**
 * Search for art only (DeviantArt, Tumblr)
 * @param {string} query - Search query
 * @param {number} maxResults - Maximum results
 * @returns {Promise<{ok: true, data: DiscoveryResult} | {ok: false, error: Object}>}
 */
export async function searchArt(query: string, maxResults = 10): Promise<any> {
  return orchestratedSearch(query, {
    platforms: ['deviantart', 'tumblr'],
    contentTypes: ['art'],
    maxResults
  });
}

/**
 * Check content discovery service health
 * @returns {Promise<{ok: true, data: Object} | {ok: false, error: Object}>}
 */
export async function checkHealth() {
  try {
    const res = await httpPostJson('/content-discovery/health', {}, AI_SERVICE_TIMEOUT_MS);

    if (res.status >= 400 || !res.json?.ok) {
      return {
        ok: false,
        error: {
          code: 'DEPENDENCY_UNAVAILABLE',
          message: 'Content discovery health check failed'
        }
      };
    }

    return {
      ok: true,
      data: res.json
    };
  } catch (err: any) {
    logger.error('[CONTENT-DISCOVERY] Health check error', { error: err.message });
    return {
      ok: false,
      error: {
        code: 'DEPENDENCY_UNAVAILABLE',
        message: 'Content discovery service unavailable',
        cause: err
      }
    };
  }
}

File: ai-service/app/services/agent/content_discovery.py
"""
Content Discovery Orchestrator

Main orchestration logic for multi-platform content discovery.
Coordinates search, scoring, and ranking across all platforms.
"""

import asyncio
from typing import List, Dict, Any, Optional
from .relevance_scorer import RelevanceScorer
from .diversity_ranker import DiversityRanker


class ContentDiscoveryOrchestrator:
    """Orchestrate content discovery across multiple platforms"""

    def __init__(
        self,
        deviantart_client=None,
        tumblr_client=None,
        web_search_func=None,
        llm_service=None,
    ):
        """
        Initialize orchestrator with platform clients.

        Args:
            deviantart_client: DeviantArt API client
            tumblr_client: Tumblr API client
            web_search_func: Web search function (for news, YouTube, Reddit, Twitter)
            llm_service: LLM service for relevance scoring
        """
        self.deviantart = deviantart_client
        self.tumblr = tumblr_client
        self.web_search = web_search_func

        # Initialize scorer and ranker
        self.scorer = RelevanceScorer(llm_service=llm_service)
        self.ranker = DiversityRanker()

    async def discover(
        self,
        query: str,
        platforms: List[str] = None,
        max_results: int = 20,
        content_types: List[str] = None,
    ) -> Dict[str, Any]:
        """
        Main orchestration method for content discovery.

        Executes multi-stage pipeline:
        1. Stage 1: Multi-source aggregation (parallel)
        2. Stage 2: LLM-powered relevance scoring (batch)
        3. Stage 3: Diversity-aware ranking

        Args:
            query: Search query
            platforms: List of platforms to search (default: all)
            max_results: Maximum total results to return
            content_types: Desired content types (default: ['news', 'video', 'discussion', 'art'])

        Returns:
            Discovery results with scored and ranked items
        """
        if platforms is None:
            platforms = ["news", "youtube", "reddit", "twitter", "deviantart", "tumblr"]

        if content_types is None:
            content_types = ["news", "video", "discussion", "art"]

        # Stage 1: Parallel Multi-Source Aggregation
        raw_results = await self._aggregate_sources(query, platforms, max_results * 2)

        if not raw_results:
            return {
                "query": query,
                "results": [],
                "total_results": 0,
                "diversity_score": 0.0,
                "platforms_covered": [],
            }

        # Stage 2: LLM-Powered Relevance Scoring
        scored_results = await self.scorer.score_batch(raw_results, query)

        # Stage 3: Content Diversity & Ranking
        final_results = self.ranker.rank(scored_results, content_types)

        # Calculate diversity score
        diversity_score = self.ranker.calculate_diversity(final_results)

        # Get platforms that returned results
        platforms_covered = list(set(r.get("platform", "") for r in final_results))

        return {
            "query": query,
            "results": final_results[:max_results],
            "total_results": len(final_results),
            "diversity_score": diversity_score,
            "platforms_covered": platforms_covered,
        }

    async def _aggregate_sources(
        self,
        query: str,
        platforms: List[str],
        max_per_platform: int = 10,
    ) -> List[Dict[str, Any]]:
        """
        Stage 1: Aggregate results from all enabled platforms in parallel.

        Args:
            query: Search query
            platforms: List of platforms to search
            max_per_platform: Maximum results per platform

        Returns:
            Combined list of results from all platforms
        """
        tasks = []

        # Queue up all platform searches
        if "news" in platforms:
            tasks.append(self._search_news(query, max_per_platform))

        if "youtube" in platforms:
            tasks.append(self._search_youtube(query, max_per_platform))

        if "reddit" in platforms:
            tasks.append(self._search_reddit(query, max_per_platform))

        if "twitter" in platforms:
            tasks.append(self._search_twitter(query, max_per_platform))

        if "deviantart" in platforms and self.deviantart:
            tasks.append(self._search_deviantart(query, max_per_platform))

        if "tumblr" in platforms and self.tumblr:
            tasks.append(self._search_tumblr(query, max_per_platform))

        # Execute all searches in parallel
        results_by_platform = await asyncio.gather(*tasks, return_exceptions=True)

        # Flatten and deduplicate
        all_results = []
        seen_urls = set()

        for platform_results in results_by_platform:
            if isinstance(platform_results, Exception):
                # Log error but continue with other platforms
                continue

            if isinstance(platform_results, list):
                for item in platform_results:
                    url = item.get("url", "")
                    if url and url not in seen_urls:
                        seen_urls.add(url)
                        all_results.append(item)

        return all_results

    async def _search_news(self, query: str, max_results: int) -> List[Dict]:
        """Search news sites using Brave API"""
        if not self.web_search:
            return []

        try:
            results = await self.web_search(
                query=query,
                max_results=max_results,
                domains=[
                    "variety.com",
                    "hollywoodreporter.com",
                    "deadline.com",
                    "ew.com",
                    "ign.com",
                ],
                recency_days=7,
            )

            # Normalize to common schema
            return [self._normalize_web_result(r, "news") for r in results]

        except Exception:
            return []

    async def _search_youtube(self, query: str, max_results: int) -> List[Dict]:
        """Search YouTube using Brave API"""
        if not self.web_search:
            return []

        try:
            results = await self.web_search(
                query=query, max_results=max_results, domains=["youtube.com"], recency_days=30
            )

            return [self._normalize_web_result(r, "video") for r in results]

        except Exception:
            return []

    async def _search_reddit(self, query: str, max_results: int) -> List[Dict]:
        """Search Reddit using Brave API"""
        if not self.web_search:
            return []

        try:
            results = await self.web_search(
                query=query, max_results=max_results, domains=["reddit.com"], recency_days=7
            )

            return [self._normalize_web_result(r, "discussion") for r in results]

        except Exception:
            return []

    async def _search_twitter(self, query: str, max_results: int) -> List[Dict]:
        """Search Twitter/X using Brave API"""
        if not self.web_search:
            return []

        try:
            results = await self.web_search(
                query=query,
                max_results=max_results,
                domains=["twitter.com", "x.com"],
                recency_days=3,
            )

            return [self._normalize_web_result(r, "discussion") for r in results]

        except Exception:
            return []

    async def _search_deviantart(self, query: str, max_results: int) -> List[Dict]:
        """Search DeviantArt for fan art"""
        if not self.deviantart:
            return []

        try:
            results = await self.deviantart.search_by_tag(
                query=query, max_results=max_results, mature_content=False
            )

            # Results already normalized by DeviantArt client
            return results

        except Exception:
            return []

    async def _search_tumblr(self, query: str, max_results: int) -> List[Dict]:
        """Search Tumblr for posts"""
        if not self.tumblr:
            return []

        try:
            results = await self.tumblr.search_by_tag(
                query=query, max_results=max_results, filter_nsfw=True
            )

            # Results already normalized by Tumblr client
            return results

        except Exception:
            return []

    def _normalize_web_result(self, result: Dict, content_type: str) -> Dict[str, Any]:
        """
        Normalize Brave API results to common schema.

        Args:
            result: Raw result from Brave API
            content_type: Content type (news, video, discussion)

        Returns:
            Normalized result dictionary
        """
        return {
            "title": result.get("title", ""),
            "url": result.get("url", ""),
            "snippet": result.get("snippet", "") or result.get("description", ""),
            "platform": result.get("domain", "unknown"),
            "content_type": content_type,
            "published_date": result.get("published_date"),
            "description": result.get("snippet", ""),
            # No attribution for web results (not art)
            "attribution": None,
        }
